### PR TITLE
Fix update_model_requirements

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/13519?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13519/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13519
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Fix update_model_requirements for List[str]

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

Added API doc
image

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by harupy** [comment]
why do we need this PR?

**Comment by serena-ruan** [comment]
> why do we need this PR?

@harupy Because `update_model_requirements` is used inside `update_pip_requirements` where requirements are List[Requirement], but if users directly use `update_model_requirements` we're expecting it to be List[str]. Inside `_remove_requirements` we expect List[Requirement] again, so we need to convert them to Requirement

**Comment by harupy** [comment]
can we fix update_pip_requirements instead?

**Comment by serena-ruan** [comment]
> can we fix update_pip_requirements instead?

`update_pip_requirements` is working fine. But do we expect users to call `update_pip_requirements` instead of `update_model_requirements`?

**Comment by harupy** [comment]
The issue is `update_pip_requirements` calling `update_model_requirements` with `Requirement`, right? Can we fix this instead?

File: docs/source/python_api/mlflow.models.rst [hunk 1]
```diff
@@ -8,3 +8,5 @@ mlflow.models
 
 .. autoclass:: mlflow.models.model.ModelInfo
     :members:
+
+.. autofunction:: mlflow.models.model.update_model_requirements
```
Comment (harupy) [inline docs/source/python_api/mlflow.models.rst:12]: Is `update_model_requirements` imported in `mlflow.models`?

File: mlflow/models/model.py [hunk 2]
```diff
@@ -996,6 +996,12 @@ def update_model_requirements(
     old_conda_reqs = _get_requirements_from_file(conda_yaml_path)
     old_requirements_reqs = _get_requirements_from_file(requirements_txt_path)
 
+    try:
+        requirement_list = [Requirement(s.strip().lower()) for s in requirement_list]
+    except Exception as e:
+        raise MlflowException.invalid_parameter_value(
+            f"Invalid requirement list: {requirement_list}"
+        ) from e
```
Comment (harupy) [inline mlflow/models/model.py:1004]: /ai make two changes

1. Replace `Exception` with `InvalidRequirement`
2. Only inclue invalid requirements in the error message.

File: docs/source/python_api/mlflow.models.rst (hunk 1 above)
Comment (serena-ruan) [reply docs/source/python_api/mlflow.models.rst:12]: It's not

Comment (harupy) [reply docs/source/python_api/mlflow.models.rst:12]: can you import it?

File: mlflow/models/model.py [hunk 3]
```diff
@@ -996,14 +996,25 @@ def update_model_requirements(
     old_conda_reqs = _get_requirements_from_file(conda_yaml_path)
     old_requirements_reqs = _get_requirements_from_file(requirements_txt_path)
 
+    requirements = []
+    invalid_requirements = {}
+    for s in requirement_list:
+        try:
+            requirements.append(Requirement(s.strip().lower()))
+        except InvalidRequirement as e:
+            invalid_requirements[s] = e
```
Comment (harupy) [inline mlflow/models/model.py:1005]: What does the error message look like? If it just says "invalid requirement", we don't need to include it in the message.

File: mlflow/models/model.py [hunk 4]
```diff
@@ -996,14 +996,25 @@ def update_model_requirements(
...
+    if invalid_requirements:
+        raise MlflowException.invalid_parameter_value(
+            f"Invalid requirement list: {invalid_requirements}"
```
Comment (harupy) [inline mlflow/models/model.py:1008]:
```suggestion
            f"Found requirements: {invalid_requirements}"
```

File: docs/source/python_api/mlflow.models.rst (hunk 1 above)
Comment (harupy) [reply docs/source/python_api/mlflow.models.rst:12]: Never mind if `mlflow.models.model` is already public.

File: tests/models/test_cli.py [hunk 5]
```diff
@@ -950,3 +950,43 @@ def test_update_requirements_cli_throws_on_incompatible_input():
...
+    update_model_requirements(
+        model_uri, "add", ["mlflow>=2.9, !=2.9.0", "coolpackage[extra]==8.8.8"]
+    )
+
+    resolved_uri = RunsArtifactRepository.get_underlying_uri(model_uri)
+    local_paths = get_model_requirements_files(resolved_uri)
+
+    # the tool should overwrite mlflow, add coolpackage, and leave cloudpickle alone
+    reqs = _get_requirements_from_file(local_paths.requirements)
+    assert Requirement("mlflow!=2.9.0,>=2.9") in reqs
```
[the remark this answers is not in the document]
Comment (serena-ruan) [reply tests/models/test_cli.py:968]: local_paths.requirements is a Path, _get_requirements_from_file convert it into a Requirement object

File: mlflow/models/model.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/models/model.py:1005]:
```
>>> invalid_reqs = ["mlflow=", "abc>", "~package"]
>>> invalid_requirements = {}
>>> for req in invalid_reqs:
...     try:
...             Requirement(req)
...     except Exception as e:
...             invalid_requirements[req] = e
... 
>>> invalid_requirements
{'mlflow=': InvalidRequirement('Expected end or semicolon (after name and no valid version specifier)\n    mlflow=\n          ^'), 'abc>': InvalidRequirement('Expected end or semicolon (after name and no valid version specifier)\n    abc>\n       ^'), '~package': InvalidRequirement('Expected package name at the start of dependency specifier\n    ~package\n    ^')}
```
The error message seems useful

**Review by harupy — APPROVED** [review]
LGTM！

**Comment by serena-ruan** [comment]
Failing recipes test not related, merging this PR

**Pull request merged**